#### Imbalanced Dataset Handling With Machine Learning 

In [1]:
import pandas as pd 

In [2]:
df = pd.read_csv(r'C:\Users\User\Desktop\Github_Upload_datasets\creditcard.csv')
df.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

In [3]:
df.shape

(284807, 31)

In [4]:
df.columns

Index(['Time', 'V1', 'V2', 'V3', 'V4', 'V5', 'V6', 'V7', 'V8', 'V9', 'V10',
       'V11', 'V12', 'V13', 'V14', 'V15', 'V16', 'V17', 'V18', 'V19', 'V20',
       'V21', 'V22', 'V23', 'V24', 'V25', 'V26', 'V27', 'V28', 'Amount',
       'Class'],
      dtype='object')

In [5]:
df.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

In [6]:
df.count()

Time      284807
V1        284807
V2        284807
V3        284807
V4        284807
V5        284807
V6        284807
V7        284807
V8        284807
V9        284807
V10       284807
V11       284807
V12       284807
V13       284807
V14       284807
V15       284807
V16       284807
V17       284807
V18       284807
V19       284807
V20       284807
V21       284807
V22       284807
V23       284807
V24       284807
V25       284807
V26       284807
V27       284807
V28       284807
Amount    284807
Class     284807
dtype: int64

From above it is clear that there is null value as every feature has same counts as well as isnull().sum() is also zero.

We will check dependent feature output value ...

In [7]:
df['Class']

0         0
1         0
2         0
3         0
4         0
         ..
284802    0
284803    0
284804    0
284805    0
284806    0
Name: Class, Length: 284807, dtype: int64

This is a classification problem that is whether a transaction is fraud or not. for this we will be taking class features. we will check how many transactions are fraud (zero) and how many are not that is one. Especially checking number of Zeros and One.

In [8]:
df['Class'].value_counts()

0    284315
1       492
Name: Class, dtype: int64

Therefor, Number of Zeros(Fraud Detection) are 284315 and Number of Ones(Genuine Transaction) are 492.

Huge difference between zeros and ones. 

#### How do you know your dataset is imbalaced or not  ?? 

Here above, you are seeing that number of Zeros is far more that number of ones and for an example, suppose you have number of ones = 100 and number of zeros = 25,that only 25%. through this way we will undersatnd that our dataset is imbalanced.

Classification predictive modelling involves ppredicting a class label for a given observation.

An imbalanced classification problem is an example of a classification problem where the distribution of examples across the known classes is biased or skewed.

Classification is a predictive modeling problem that involves assigning a class label to each observation.
....... classification models generate a predicted class, which comes in the form of a discrete category. for most pratical applications, a discrete category(0 or 1) prediction is required in order to make a decision.

manyreal-world classification problems have an imbalance class distribution, such as fraud detection, spam detection, and churn prediction.

Imbalanced Dataset: Any dataset with an unequal class distribution is technically imbalanced. However, a dataset is said to be imbalanced when there is a significant, or in some cases extreme, disproportion among the number of examples of each class of the problem.

A slight imbalance is often not a concern, and the problem can often be treated like a normal problem can often be treated like a normal classification predictive modeling problem.
A severe imbalance of the classes can be challenging to model and may require the use of specialized techniques.

" Any dataset with an unequal class distribution is technically imbalanced.However, a dataset is said to be imbalanced when there is a significant, or in some cases extreme, disproportion among the number of examples of each class of the problem.

In [9]:
The imbalance of the class distribution will vary across problems.
A classification problem may be a little skewed, such as if there is a slight imbalance. Alternatively, the classification problem may have a severe imbalance where there might be hundreds or thousands of examples in one class and tens of examples in another class for a given training dataset.

* Slight imbalance : An imbalance classification problem where the distribution of examples is uneven by a small amount in the training dataset (e.g. 4:6).
* Severe imbalance : An imbalance classification problem where the distribution of examples is uneven by a large amount in the training dataset (e.g 1:100 or more)

    "Most of the contemporary works in class imbalance concentrate on imblance ratios ranging from 1:4 up to 1:100. In real-life applications such as fraud detection or cheminformatics we may deal with problems with imbalance ratio ranging from 1:1000 up to 1:5000".

SyntaxError: invalid syntax (<ipython-input-9-63060062b84f>, line 1)

In [ ]:
# Independent and Dependent Features
X = df.drop('Class', axis = 1)
y = df.Class 

Imbalanced dataset is not heavly impacted by ensemble technique or where decision tree works.let's see....
we will use basic classification model first and then compare, see the result.

#### Cross Validation Like KFold and Hyperparameter Tuning

In [ ]:
from sklearn.linear_model import LogisticRegression 
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report
from sklearn.model_selection import KFold
import numpy as np 
from sklearn.model_selection import GridSearchCV

In [ ]:
10.0**np.arange(-2,3)

In [ ]:
log_class = LogisticRegression()
grid={'C':10.0**np.arange(-2,3), 'penalty':['l1', 'l2']}
cv = KFold(n_splits = 5, random_state = None, shuffle = False)

In [ ]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X,y, train_size = 0.7)

In [ ]:
clf = GridSearchCV(log_class, grid, cv=cv, n_jobs = -1, scoring='f1_macro')
clf.fit(X_train, y_train)

In [ ]:
y_pred = clf.predict(X_test)
print(confusion_matrix(y_test, y_pred))
print(accuracy_score(y_test,y_pred))
print(classification_report(y_test, y_pred))

You should not focus on accuracy(0.99), we should focus on how to increase precision (0.83) and recall values(0.74). how you will decrease the false positive value(22) and false negative value(40).

Let's use some other models too ....

In [ ]:
y_train.value_counts()

In [ ]:
341*100

In [ ]:
class_weight = dict({0:1, 1:100})

In [ ]:
from sklearn.ensemble import RandomForestClassifier 
classifier = RandomForestClassifier(class_weight = class_weight)
classifier.fit(X_train, y_train)

let's understand class_weight parameter first . 
supppose you have zero records 10000 and one records 100 . now you have to tell random forest to give more imprtance to ones ..?? how will you do that ..?
you will create a balanced condition that is 1:1 but through class_weight what we have done is we put class_weight = 1:100 , that 100 times more imprtance to 1 rather 0.

        0        1
        10000    100
        
        1:100 
            
 probably we will get good results ...class_weight parameters will be assign to this model.           

In [ ]:
y_pred = classifier.predict(X_test)
print(confusion_matrix(y_test, y_pred))
print(accuracy_score(y_test,y_pred))
print(classification_report(y_test, y_pred))

You should not focus on accuracy(0.99), we should focus on how to increase precision (0.97) and recall values(0.77). how you will decrease the false positive value(3) and false negative value(34).  

After cross validation and hyperparameter tuning in RandomForest, you will get much more good results. people usually have favourite models like RandomForest, XGboost, AdaBoost. here XGboost will perform very well here. 

#### Under Sampling 

what is under sampling ?? first understand this suppose you have zeros = 1000 and ones = 100 , then in under sampling we will reduce the points of the maximum labels. i.e we will 1000 labels to 200 labels, such sampling is not preferred for the large datasets as this causes Data Loss. preffered only for small datasets....

In [ ]:
y_train.value_counts()

In [ ]:
from collections import Counter
Counter(y_train)

In [ ]:
from imblearn.under_sampling import NearMiss
ns = NearMiss(0.8) # reduce it upto 80%
X_train_ns, y_train_ns = ns.fit_resample(X_train, y_train)
print('The number of classes before fit {}'.format(Counter(y_train)))
print('The number of classes after fit {}'.format(Counter(y_train_ns)))


In [ ]:
0.8*426

In [ ]:
from sklearn.ensemble import RandomForestClassifier 
classifier = RandomForestClassifier()
classifier.fit(X_train_ns, y_train_ns)

In [ ]:
y_pred = classifier.predict(X_test)
print(confusion_matrix(y_test, y_pred))
print(accuracy_score(y_test,y_pred))
print(classification_report(y_test, y_pred))

Here, precision is very low (0.01) and False positive values that is  increased to 17567 which means so many errors. Never do under-sampling under the dataset is very very small. 

#### Over Sampling 

In [ ]:
from imblearn.over_sampling import RandomOverSampler

In [ ]:
os = RandomOverSampler(0.75) 
X_train_ns, y_train_ns = os.fit_resample(X_train, y_train)
print('The number of classes before fit {}'.format(Counter(y_train)))
print('The number of classes after fit {}'.format(Counter(y_train_ns)))

In over sampling what we do that in over sampling we generally keep increasing those values , that is from 341 to 99511 .

In [ ]:
from sklearn.ensemble import RandomForestClassifier 
classifier = RandomForestClassifier()
classifier.fit(X_train_ns, y_train_ns)

In [ ]:
y_pred = classifier.predict(X_test)
print(confusion_matrix(y_test, y_pred))
print(accuracy_score(y_test,y_pred))
print(classification_report(y_test, y_pred))

From doing the over sampling, we have reduced the false negative too 32 which is awesome ....

Morale of the story: Never do under-sampling as you will lose lots of data.

#### Interview Question : How will you handle imbalance dataset ..... ??

1. if the dataset is small, will definetly go with under sampling but again i will focus on all the performance matrix like recall, precision, f1_score apart from that i will be also be focussing based on the domain knowledge whether i should be reducing the false positive with the false negative based on that i will selecting roc score that is the probability value whether it should be 0.6 0r 0.7 something like that.
2. finally i will use more techniques like doing over sampling but again will be always involving around in understanding the performance metrics.
3. finally if this will not work then i will go for ensemble technique like random forest, decision tree , xgboost , adaboost where i can provide my class_weight parameters.

#### SMOTETomek

Uses the concept of Under Sampling and Over Sampling 

In [ ]:
from imblearn.combine import SMOTETomek

In [ ]:
os = SMOTETomek(0.75) 
X_train_ns, y_train_ns = os.fit_resample(X_train, y_train)
print('The number of classes before fit {}'.format(Counter(y_train)))
print('The number of classes after fit {}'.format(Counter(y_train_ns)))

In [ ]:
from sklearn.ensemble import RandomForestClassifier 
classifier = RandomForestClassifier()
classifier.fit(X_train_ns, y_train_ns)

In [ ]:
y_pred = classifier.predict(X_test)
print(confusion_matrix(y_test, y_pred))
print(accuracy_score(y_test,y_pred))
print(classification_report(y_test, y_pred))

When to use over sampling and under sampling ??
it should be done during the model creation, always remember. Don't confuse like before the EDA like that.

#### Ensemble Techniques 

In [ ]:
from imblearn.ensemble import EasyEnsembleClassifier

In [ ]:
easy = EasyEnsembleClassifier()
easy.fit(X_train, y_train)

In [ ]:
y_pred = easy.predict(X_test)
print(confusion_matrix(y_test, y_pred))
print(accuracy_score(y_test,y_pred))
print(classification_report(y_test, y_pred))